### PR TITLE
AB#8726 Rework APK versioning

### PR DESCRIPTION
**Pull request by chpa7**

### CODE REVIEW

File: play-media-native/package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "play-media-native",
-  "version": "1.0.0",
```
Comment (jflheureux) [inline play-media-native/package.json:3]: I think we still need the version attribute.
https://docs.npmjs.com/cli/v9/configuring-npm/package-json#version

File: play-media-native/scripts/publishAPKGitHub.mts [hunk 2]
```diff
@@ -19,40 +20,26 @@ const updateTagVersion = (tag: string): string => {
...
-    (res: { data: any[] }) =>
-      res.data.find((release: { prerelease: boolean }) => release.prerelease === true).tag_name
-  );
-} else {
-  tagName = await request('GET /repos/{owner}/{repo}/releases/latest', {
-    ...defaultOptions,
-    headers: defaultHeaders,
-  }).then((res: { data: { tag_name: string } }) => res.data.tag_name);
-}
+// Retrieve the latest nightly tag name and update it
```
Comment (jflheureux) [inline play-media-native/scripts/publishAPKGitHub.mts:35]: Thinking about it. It could happen that we do 2 successive stable builds without a nightly in between. I think we should get both the latest nightly and the latest stable releases that are matching the baseVersion. Then take the highest build number, then adding +1. Same when releasing a nightly package.

**Review by jflheureux — CHANGES_REQUESTED** [review]

File: play-media-native/package.json (hunk 1 above)
Comment (chpa7) [reply play-media-native/package.json:3]: Version attribute is optional for our case (mandatory only when publishing to npm). We could put it back but I think it will be kinda confusing having both baseVersion and version in there.

File: play-media-native/scripts/publishAPKGitHub.mts [hunk 3]
```diff
@@ -15,44 +16,50 @@ const defaultHeaders = {
...
-    ...defaultOptions,
-    headers: defaultHeaders,
-  }).then((res: { data: { tag_name: string } }) => res.data.tag_name);
-}
+// Retrieve the latest nightly and stable tag name matching the base version and update the one with the highest build number
+const tagName = await request('GET /repos/{owner}/{repo}/releases', {
+  ...defaultOptions,
+  headers: defaultHeaders,
+}).then((res: { data: any[] }) => {
+  const latestNightlyTagName = res.data.find(
```
Comment (jflheureux) [inline play-media-native/scripts/publishAPKGitHub.mts:44]: I think this logic will not work if we bump the baseVersion minor or major version and we have no existing release with that base version. Both `latestNightlyTagName` and `latestStableTagName` will be null.

Also, if GitHub returns the releases in an order different from created date descending, we could get in trouble by returning the first one in the array that matches the baseVersion, even if it is not the highest build number.

I created Jest unit tests and improved the logic to have a single function that returns the tag with the absolute highest build number that matches the base version. I also return undefined when none exist. And I improved the logic that follow to handle the case when it is undefined.

**Review by jflheureux — APPROVED** [review]
As this was complex code, I wrote unit tests to validate my assumptions and test some corner cases for me. While being in the code, I took the time to update the code to cover these corner cases. Let me know if you do not like when I do this in your PRs.
